mirrord-layer/src/files.rs

```diff
@@ -0,0 +1,194 @@
+/// (which is different anyways when running in `-agent` context).
+///
+/// When called for a valid file, it'll block, send a `ClientMessage::OpenFileRequest` to be handled
+/// by `mirrord-agent` (`handle_peer_message` function), and wait until it receives a
+/// `DaemonMessage::FileOpenResponse`.
+pub(crate) fn open(path: PathBuf) -> Result<RawFd, LayerError> {
+    debug!("open -> trying to open valid file {path:?}.");
+
+    let sender = unsafe { HOOK_SENDER.as_ref().unwrap() };
+    let (file_channel_tx, file_channel_rx) = oneshot::channel::<OpenFileResponse>();
+
+    let requesting_file = OpenFileHook {
+        path,
+        file_channel_tx,
+    };
+
+    sender.blocking_send(HookMessage::OpenFileHook(requesting_file))?;
+
+    debug!("Blocking on file open operation!");
+    let open_file = file_channel_rx.blocking_recv()?;
+    debug!("After `block_on` call, we have an open file {open_file:#?}!");
+
+    OPEN_FILES.lock().unwrap().insert(open_file.fd);
+
+    Ok(open_file.fd)
+}
+
+/// Blocking wrapper around `libc::read` call.
+///
+/// Bypassed when trying to load system files, and files from the current working directory, see
+/// `open`.
+pub(crate) fn read(fd: RawFd, read_amount: usize) -> Result<ReadFile, LayerError> {
+    debug!("read -> trying to read valid file {fd:?}.");
+
+    let sender = unsafe { HOOK_SENDER.as_ref().unwrap() };
+    let (file_channel_tx, file_channel_rx) = oneshot::channel::<ReadFileResponse>();
+
+    let read_file = ReadFileHook {
+        fd,
+        buffer_size: read_amount,
+        file_channel_tx,
+    };
+
+    sender.blocking_send(HookMessage::ReadFileHook(read_file))?;
+
+    let read_file_response = file_channel_rx.blocking_recv()?;
+
+    let read_file = ReadFile {
+        bytes: read_file_response.bytes,
+        read_amount: read_file_response.read_amount,
+    };
+
+    Ok(read_file)
+}
+
+/// NOTE(alex): libc also has an `open64` function. Both functions point to the same address, so
+/// trying to intercept them all will result in an `InterceptorAlreadyReplaced` error.
+pub(super) unsafe extern "C" fn open_detour(raw_path: *const c_char, oflag: c_int) -> RawFd {
+    let path: PathBuf = CStr::from_ptr(raw_path)
+        .to_str()
+        .expect("Failed converting path from c_char!")
+        .into();
+
+    if IGNORE_FILES.is_match(path.to_str().unwrap()) {
+        let fd = libc::open(raw_path, oflag);
+        fd
+    } else {
+        match open(path).map_err(|fail| {
+            error!("Failed opening file with {fail:#?}");
+            libc::EFAULT
+        }) {
+            Ok(fd) => fd,
+            Err(fail) => fail,
+        }
+    }
+}
+
+unsafe extern "C" fn read_detour(fd: c_int, buf: *mut c_void, count: size_t) -> ssize_t {
+    // NOTE(alex): We're only interested in files that are handled by `mirrord-agent`.
+    if let Some(managed_fd) = OPEN_FILES.lock().unwrap().get(&fd) {
```

Comment:
you're locking the `OPEN_FILES` mutex during the whole read, which is quite unnecessary imo.